src/lib/MapLibre.svelte

```diff
@@ -165,7 +204,33 @@
   }
 
   let lastStyle = style;
+
+  // If the last style is different from the current one
+  // we grab a list of the currrent layers and sources
+  // compare this with the stored list of layer and source ids
+  // to pick out the layers / sources which are not part of the current
+  // basemaps.
+  // We then update the style which will trigger the style.load event on
+  // map which will in turn add the user defined sources and layers back
+  // on to the map
+
   $: if ($mapInstance && !compare(style, lastStyle)) {
+    const oldMapStyle = $mapInstance.getStyle();
+
+    if (lastStyleLayerIds) {
+      layersToReAddAfterStyleChange = oldMapStyle.layers.filter(
+        (l) => !lastStyleLayerIds!.includes(l.id)
+      );
+    }
+    if (lastStyleSourceIds) {
+      const nonStyleSourceIds = Object.keys(oldMapStyle.sources).filter(
+        (sourceId) => !lastStyleSourceIds!.includes(sourceId)
+      );
+      sourcesToReAddAfterStyleChange = {};
+      for (const id of nonStyleSourceIds) {
+        sourcesToReAddAfterStyleChange[id] = oldMapStyle.sources[id];
```

Comment:
I was thinking about cases where the `paint` or `layout` prop of a `Layer` component gets modified. Doing the work here based on `map.getStyle()` is nice, because it'll pick up the current state before changing style.

One sanity check though -- we don't need to make defensive copies with `structuredClone` or similar here? I don't _think_ so, because `map.getStyle()` will return new things after the `setStyle` below, and the user-managed sources/layers won't be there. The maplibre API isn't terribly clear about ownership/lifetime of source/layer specifications...

Reply:
I dont think so either. Especially as the only thing we are storing from this entire function call are the layersToReAddAfterStyleChange and  sourcesToReAddAfterStyleChange, objects which are newly constructed each time. We can throw in a structuredClone just in case if we want to be 100% sure. The style shouldn't change often enough that we need to worry about keeping this code too performant. Miight be better to be defensive?